Handle linting

thompsonmj opened the pull request:
Run ruff; remove unused imports; add ruff and pre-commit to dev dependencies; add linter pre-commit hook; update README for developers; add lint check to GitHub workflow

This PR integrates `ruff check .` to enforce lightweight, non-opinionated linting.
Focus is on code hygiene without dictating style. 
Complimentary to `pytest` , catching issues including:

- Unused imports and variables
- Leftover debugging statements (`print`, `breakpoint`)
- Redundant or unreachable code

Only `ruff check` (not `ruff format`) is used in the pre-commit hook and CI here.

thompsonmj committed:
Handle linting

Run ruff; remove unused imports; add ruff and pre-commit to dev dependencies; add linter pre-commit hook; update README for developers; add lint check to GitHub workflow

thompsonmj committed:
Merge branch 'main' into linting

egrace479 commented on README.md:
````diff
@@ -160,3 +160,21 @@ file_paths = gather_file_paths(input_path, ignore_file=ignore_file)
+1. Clone the repository and create a branch
+2. Install with dev dependencies:
+```bash
+pip install -e .[dev]
````
```suggestion
pip install -e ".[dev]"
```
For general and `zsh` compatibility.

egrace479 reviewed (commented):
One small fix

thompsonmj committed:
Cross-platform dev installation quotations

For greatest cross-platform compatibility, suggest using double-quotes with pip install -e ".[dev]"

egrace479 reviewed (approved):
Looks good! I like the updates with `ruff` to find extraneous imports.
